prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -540,35 +541,52 @@ group x = case changesUponFlattening x of
 -- See https://github.com/quchen/prettyprinter/issues/22 for the  corresponding
 -- ticket.
 
+data FlattenResult a
+    = Flattened a
+    -- ^ a is likely flatter than the input.
+    | AlreadyFlat
+    -- ^ The input was already flat, e.g. a 'Text'.
+    | NeverFlat
+    -- ^ The input couldn't be flattened: It contained a 'Line' or 'Fail'.
```

Comment:
The main idea of this PR is to represent the old `Nothing` result with _two_ constructors: One to indicate that the result is already flat enough, the other to indicate that flattening is impossible, and which allows shortcuts.